### PR TITLE
Markup for logging, including explanations

### LINKED ISSUE #42: Logging points for testing
This is not a fully baked issue, but more a matter for discussion: testing implementations will be fairly difficult. A simple comparison of the end result would not help too much. I wonder whether it is possible to build into the algorithms some "logging points": implementations would be welcome to log some specific values at those points in the code, comparing with log result of other implementations, thereby helping to locate possible bugs more quickly. The combination of this with the test files themselves may be helpful.

Is this a viable thing to do?

### PR DESCRIPTION
**Pull request by gkellogg**
... using ~~existing and future~~ examples.

This includes examples from two test cases, ~~which haven't yet been added to the spec~~.

* double circle from [test022](https://w3c.github.io/rdf-canon/tests/#manifest-urdna2015#test022)
* duplicate paths from [test035](https://w3c.github.io/rdf-canon/tests/#manifest-urdna2015#test035)

Note that I added entries to the "How to Read This Document" section, which likely won't show up in the preview.

<dl>
  <dt><span class="details"><span class="summary">Explanation</span></span></dt><dd>
    An expandable area to find a more details non-normative explanation of a
    particular algorithmic step.
    <details>
      <summary>Explanation</summary>
      <p>This area would provide more information about the step involved.</p>
    </details>
  </dd>
  <dt><span class="details"><span class="summary">Logging</span></span></dt><dd>
    An expandable area to find suggestions for implementations to log
    information about processing,
    which may be useful in comparing with other implementations,
    or with logs provided with each test case.
    <details>
      <summary>Logging</summary>
      <p>For example, the following output snippet might
        describe the operation of an implementation using the [[?YAML]] format.</p>
      <pre class="logging yaml" data-transform="updateExample">
        
      </pre>
    </details>
  </dd>
</dl>

Fixes #42.

***

### :boom: Error: 503 Service Unavailable :boom: ###

[PR Preview](https://github.com/tobie/pr-preview#pr-preview) failed to build. _(Last tried on Jan 14, 2023, 10:57 PM UTC)_.

<details>
<summary>More</summary>

PR Preview relies on a number of web services to run. There seems to be an issue with the following one:

:rotating_light: [Spec Generator](https://www.w3.org/2015/labs/) - Spec Generator is the web service used to build specs that rely on ReSpec.

:link: [Related URL](https://labs.w3.org/spec-generator/?type=respec&url=https%3A%2F%2Fraw.githubusercontent.com%2Fw3c%2Frdf-canon%2Fb8bc0ab8b6d8f205008f74ed9656b232543eb902%2Fspec%2Findex.html%3FisPreview%3Dtrue)

```
<html><body><h1>503 Service Unavailable</h1>
No server is available to handle this request.
</body></html>

```

_If you don't have enough information above to solve the error by yourself (or to understand to which web service the error is related to, if any), please [file an issue](https://github.com/tobie/pr-preview/issues/new?title=Error%20not%20surfaced%20properly&body=See%20w3c/rdf-canon%2363.)._
</details>

### CODE REVIEW

**Comment by gkellogg** [comment]
I also note that the _pre_ blocks including the log examples are empty, although they are in the source and show yup when rendered locally. To see the full intent, run from the branch off your own local server.

**Comment by gkellogg** [comment]
You should be able to see the logging markup properly rendered here: https://raw.githack.com/w3c/rdf-canon/logging/spec/index.html#canon-algo-algo

File: spec/index.html [hunk 1]
```diff
@@ -934,6 +940,25 @@ <h3>Algorithm</h3>
               </details>
             </li>
           </ol>
+          <details>
+            <summary>Logging</summary>
+            <p>Log the sate of the <a>blank node to quads map</a>:</p>
```
Comment (TallTed) [inline spec/index.html:945]:
```suggestion
            <p>Log the state of the <a>blank node to quads map</a>:</p>
```

File: spec/common/typographical-conventions.html [hunk 2]
```diff
@@ -31,6 +31,41 @@
   <dt>[<a href="">reference</a>]</dt><dd>
     A document reference (normative or informative) is enclosed in square brackets
     and links to the references section.</dd>
+  <dt><span class="details"><span class="summary">Explanation</span></span></dt><dd>
+    An expandable area to find a more details non-normative explanation of a
+    particular algorithmic step.
```
Comment (dlongley) [inline spec/common/typographical-conventions.html:36]:
```suggestion
    An expandable area to find a more detailed, non-normative explanation of a
    particular algorithmic step.
```

**Review by dlongley — APPROVED** [review]
Thanks @gkellogg! I haven't had time to implement this kind of logging to confirm the examples values against my own implementation, but this looks really helpful.

**Comment by gkellogg** [comment]
A follow-on to this might be to augment each test with a separate file containing the execution trace.

**Comment by iherman** [comment]
I do like it.

There are minor differences between the YAML log generated by my implementation and the one in the spec, but they are cosmetic (i.e., I log when leaving the hash first degree quad function, and the example logs after the full cycle is done in the main algorithm; as a result the values for the `e0-> quads` and `e1->quads` appear as two separate log items in my case and as one with two entries in the spec). I do not think we have to "micromanage" on that level.

A minor improvement may be to make the yaml keys a bit more verbose. The spec says

```
ca:
  ca2:
    bn_to_quads:
      e0:
…
```

I generate, for the same point:

```
- log point: "[info] Entering the canonicalization function (4.5.3 (2))."
  with:
    - Bnode to quads:
        e0:
…
```

which may be a bit more readable. But that is again a minor detail.

Thanks!

**Review by iherman — APPROVED** [review]

**Comment by TallTed** [comment]
I would appreciate having the added verbosity shown in @iherman's implementation.

**Comment by gkellogg** [comment]
IMO, keeping keys simple in the YAML is better for this format. Of course, the HTML render of this could do be more verbose, at the expense of exacerbating the table width issues already present.

the key names mostly come directly from the fragment identifiers in the algorithm steps, so adding more verbose step names and hyperlinks is something we could do there.

Generally, the YAML output is highly repetitive and just sticks to inputs and outputs of different algorithm steps, so expanding these and adding context seems like a better role for the HTML generation part.

**Review by yamdan — APPROVED** [review]
Thank you @gkellogg and @iherman, this is very helpful for implementors like me who are having trouble getting an implementation to pass some tests.

**Comment by gkellogg** [comment]
The spec is now updated with examples, detailed logs, and updated details on logging each step. I'll merge in a couple of days without further comments.

**Review by dlongley — APPROVED** [review]

**Pull request merged**